examples/hal_cpu_remote_server/hal_cpu_executable.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (C) Codeplay Software Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License") with LLVM
+// Exceptions; you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/LICENSE.txt
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <signal.h>
+#include <unistd.h>
+
+#include "hal_remote/hal_server.h"
+#include "hal_remote/hal_socket_client.h"
+#include "hal_remote/hal_socket_transmitter.h"
+
+void print_usage(const char *tool_name) {
+  printf("usage: %s [-n node] port\n", tool_name);
+  printf(
+      "\tnote : node is an ip address or machine name e.g. \"127.0.0.1\" "
+      "(default) or \"localhost\"\n");
+  printf(
+      "\t       port is an integer non-zero address which will be listened "
+      "on\n");
+}
+
+hal::hal_socket_transmitter transmitter;
+
+bool process_terminated = false;
+void handle_sig(int signum) {
+  // Attempt to close down gracefully
+  transmitter.shutdown();
+  process_terminated = true;
+}
+
+// Note this should not be run as root for security reasons.
+// It will also run once per connection to avoid getting into a bad state
+// It should be rerun if desired to make more than one consecutive connection.
+int main(int argc, char **argv) {
+  std::string node = "127.0.0.1";
+  // Note this default will fail as we do not allow port 0
+  uint16_t port = 0;
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = handle_sig;
+  sigaction(SIGTERM, &action, NULL);
+
+  for (;;) {
+    switch (getopt(argc, argv, "n:h?")) {
```

Comment:
`?` is not a good option for `getopt` because the return value of `'?'` for an option `-?` is indistinguishable from the return value of `'?'` for an unrecognised option, and the option `-?` is not something we should try to support since `?` is a shell metacharacter and so unless quoted, this looks for files in the current directory whose names are `-` followed by a single character.